matrix-e2e.c

```diff
@@ -0,0 +1,1767 @@
+/**
+ * Matrix end-to-end encryption support
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02111-1301 USA
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <string.h>
+#include <sqlite3.h>
+#include "libmatrix.h"
+#include "matrix-api.h"
+#include "matrix-e2e.h"
+#include "matrix-json.h"
+#include "debug.h"
+
+/* json-glib */
+#include <json-glib/json-glib.h>
+
+#include "connection.h"
+#ifndef MATRIX_NO_E2E
+#include "olm/olm.h"
+
+struct _MatrixOlmSession;
+
+struct _MatrixE2EData {
+    OlmAccount *oa;
+    gchar *device_id;
+    gchar *curve25519_pubkey;
+    gchar *ed25519_pubkey;
+    sqlite3 *db;
+    /* Mapping from MatrixHashKeyOlm to MatrixOlmSession */
+    GHashTable *olm_session_hash;
+};
+
+#define PURPLE_CONV_E2E_STATE "e2e"
+
+/* Hung off the Purple conversation with the PURPLE_CONV_E2E_STATE */
+typedef struct _MatrixE2ERoomData {
+    /* Mapping from _MatrixHashKeyInBoundMegOlm to OlmInboundGroupSession */
+    GHashTable *megolm_sessions_inbound;
+} MatrixE2ERoomData;
+
+typedef struct _MatrixHashKeyOlm {
+    gchar *sender_key;
+    gchar *sender_id;
+} MatrixHashKeyOlm;
+
+typedef struct _MatrixOlmSession {
+    gchar *sender_key;
+    gchar *sender_id;
+    OlmSession *session;
+    sqlite3_int64 unique;
+    struct _MatrixOlmSession *next;
+} MatrixOlmSession;
+
+typedef struct _MatrixHashKeyInBoundMegOlm {
+    gchar *sender_key;
+    gchar *sender_id;
+    gchar *session_id;
+    gchar *device_id;
+} MatrixHashKeyInBoundMegOlm;
+
+/* GHashFunc for two MatrixHasKeyOlm */
+static gboolean olm_inbound_equality(gconstpointer a, gconstpointer b)
+{
+    const MatrixHashKeyOlm *hk_a;
+    const MatrixHashKeyOlm *hk_b;
+    hk_a = (const MatrixHashKeyOlm *)a;
+    hk_b = (const MatrixHashKeyOlm *)b;
+
+    return !strcmp(hk_a->sender_key, hk_b->sender_key) &&
+           !strcmp(hk_a->sender_id, hk_b->sender_id);
+}
+
+/* GHashFunc for a _MatrixHashKeyOlm */
+static guint olm_inbound_hash(gconstpointer a)
+{
+    const MatrixHashKeyOlm *hk;
+    hk = (const MatrixHashKeyOlm *)a;
+    return g_str_hash(hk->sender_key) +
+           g_str_hash(hk->sender_id);
+}
+
+static void olm_hash_key_destroy(gpointer k)
+{
+    MatrixHashKeyOlm *ok = k;
+
+    g_free(ok->sender_key);
```

Comment:
how about setting the keys to random bytes before freeing them?

Reply:
ok, I'm calling clear_mem() on sender_key now.